client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java

```diff
@@ -150,6 +150,10 @@ public void start(Codec codec, Configuration configuration, Marshaller marshalle
       pingServersIgnoreException();
    }
 
+   public Codec getNegotiatedCodec() {
```

Comment:
I don't understand why this is required. can you explain?

Reply:
Sure. Basically more `OperationsFactory` s could be created from the same `ChannelFactory`. I don't know why this is possible, see for instance:

https://github.com/infinispan/infinispan/blob/f415ee6db418375bd7a2a309603357c4a7ce345d/client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java#L626

The handshake happens at `ChannelFactory` level, so if the client and the server decided to agree on a given version, all `OperationsFactory` s using this `ChannelFactory` should use the same version as well.
Otherwise, there will be some errors.

That's the reason for the commit https://github.com/infinispan/infinispan/pull/10228/commits/836f6b6d2433bdc08b65ab029a0f9eed97c0c7d4

Reply:
still does not make sense to me. `ChannelFactory` does not interact with the `codec` and the `PingResponse` sets the `codec` after the handshake.

Reply:
The handshake sets the code to the `OperationsFactory` that is contained in the `ChannelFactory`, so from that moment any other `OperationsFactory` created from the same `ChannelFactory` should use the same possible updated codec version. If we don't do that and we create an ad admin operation after the handshake, using a v4. client and a v.3.1 server, we get errors, since the client would use the 4.0 not supported by the server.
We tested it with @jabolina yesterday evening.

Reply:
@pruivo Every time `administration()` is called it instantiates a new factory. The new factory uses the configured codec, which may be using a different version that the one negotiated by the ping. This causes the `AdminOperation` to fail. We took this approach so we can use the negotiated version in other places. 

TBH I think we could completely remove the codec placed in `RemoteCacheManager` and retrieve the correct version from the factory if needed, but this could be in another  PR.

Reply:
> TBH I think we could completely remove the codec placed in `RemoteCacheManager` and retrieve the correct version from the factory if needed, but this could be in another PR.

I think the same, but maybe it is call for the next time ;)

Reply:
it does not make any sense to create a new `OperationsFactory` every time `administration()` is invoked! Since you are changing the `ChannelFactory`, it is better to return the ` OperationsFactory` instead of the `Codec`

Reply:
The creation of a new `OperationFactory` every time `administration()` is invoked is something that was already present on the main branch. We didn't change that since I don't know this part of the code, maybe there are some security reasons, e.g. the admin operation can be performed by a special role user admin. @pruivo If we say that is safe, we could avoid to recreate the `OperationFactory` if the `ChannelFactory` already has one initialized. But it looks like a big change, for that reason we didn't do that. Let us know about it.

Reply:
Or (I was thinking) the new `OperationFactory` created from `administration()` call is supposed to be passed to another thread, that could be another reason for this instance creation, and the `OperationFactory` and I don't know if the instance is thread safe. But as I said I made the choice that had the minimal impact, since I don't know the code very well here.

Reply:
Looking at the code, I don't see a reason to not reuse the `OperationFactory`
* security/authentication -> performed by `AuthHandler` using the credentials from `Configuration`
* thread safety -> all fields are final, no issue here.

Reply:
> thread safety -> all fields are final, no issue here.

Maybe some nested fields are not.

Anyway I'm going to add a commit to avoid any new `OperationFactory` instance creation from a `ChannelFactory` if the `ChannelFactory` already has a `OperationFactory` instance.
If you say it is safe, it is OK.
Thanks

Reply:
I realized that the are other cases where an operations factory is created using an already started client factory, cases where the cache name needs to be passed, since it will be used by the new operations factory.
For these further cases so it is not possible to use the cached operations factory on client factory.
Thus, in order to use the right (negotiated) codec version I think that we cannot remove these changes.

[full discussion here](https://infinispan.zulipchat.com/#narrow/stream/118645-infinispan/topic/operations.20factory.20.2F.20channel.20factory/near/290771627)